client/client.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"github.com/myteksi/hystrix-go/hystrix"
+	"http-echo/server"
+	"http-echo/server/config"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	//"io/ioutil"
+	//"net/http"
+	"sync"
+	"time"
+)
+
+var (
+	etcdClient       server.EtcdClient
+	serviceInstances ServiceInstances
+)
+
+func init() {
+	configPath := "./service.yml"
+	config := config.Loadconfig(configPath)
+	etcdClient = server.NewEtcdClient(config.Etcd.Endpoints, config.Etcd.Timeout)
+	serviceInstances = ServiceInstances{instances: map[string]ServiceInstance{}, endpoints: make([]string, 0), next: 0}
+	WatchServiceInstances(server.EtcServicePrefix, etcdClient, &serviceInstances)
+	serviceInstances.RemoveServiceInstanceBackground()
+}
+
+type ServiceInstance struct {
+	location         string
+	enable           bool
+	timeChangeStatus time.Time
+}
+
+type ServiceInstances struct {
+	instances map[string]ServiceInstance
+	mu        sync.Mutex
+	endpoints []string
+	next      int
+}
+
+func main() {
+	time.Sleep(time.Second * 5)
+	//setting default for hystrix
+	setting := hystrix.Settings{
+		Timeout:                     10000,
+		MaxConcurrentRequests:       100,
+		ErrorPercentThreshold:       15,
+		QueueSizeRejectionThreshold: 100}
+	hystrix.Initialize(&setting)
+	circuitNamePrefix := "callEchoServer"
+
+	for {
+		url, err := serviceInstances.GetServiceInstance()
+
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+		circuitInstance := circuitNamePrefix + url
+		hystrix.Do(circuitInstance, func() error {
+			fmt.Print(" call to ", url)
+			time.Sleep(10 * time.Millisecond)
+			if err != nil {
+				return err
+			}
+			resp, err := http.Get(url)
+			if err != nil {
+				return errors.New(" error call to destination ")
+			}
+			if resp.StatusCode > 200 {
+				return errors.New(" destination service error ")
+			}
+			body, err := ioutil.ReadAll(resp.Body)
+			defer resp.Body.Close()
+			fmt.Println(" data return ", string(body))
+			return nil
+		}, func(err error) error {
+			circuit, _, errGetCircuit := hystrix.GetCircuit(circuitInstance)
+			if errGetCircuit != nil {
+				return errGetCircuit
+			}
+			if circuit.IsOpen() {
+				serviceInstances.DisableServiceInstance(url)
+			}
+			fmt.Println(err)
+			return nil
+		})
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	time.Sleep(time.Hour)
+}
+
+func WatchServiceInstances(prefix string, etcdClient server.EtcdClient, serviceInstances *ServiceInstances) {
+	go func(prefix string, etcdClient server.EtcdClient, serviceInstances *ServiceInstances) {
+		for {
+			locations, _ := etcdClient.GetServiceLocation(prefix)
+			serviceInstances.mu.Lock()
```

Comment:
we should use `defer serviceInstances.mu.Unlock()` to make sure we don't forget it 
moreover if we put it at the end, when something wrong happens at the middle of this method, `return` will be called and `Unlock()` will be missed

Reply:
My idea was using mu to lock serviceInstances during modify it's data to avoid race condition can be raise from other function. i wonder if we use `defer serviceInstances.mu.Unlock()`  this mothod need to run after  `time.Sleep(10 * time.Second)` . I think i will move block code what need lock and unlock to function so we can use  `defer serviceInstances.mu.Unlock()`.